src/wifi.rs

```diff
@@ -1442,3 +1445,144 @@ impl WifiWait {
+        self.base_driver.borrow_mut().stop()
+        // TODO: wait until stop is complete?
+    }
+
+    pub async fn connect(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().connect()
+        // TODO: wait until connect is complete
+    }
+
+    pub async fn disconnect(&mut self) -> Result<(), EspError> {
+        self.base_driver.borrow_mut().disconnect()
+        // TODO: wait until stop is complete
+    }
+
+    pub async fn scan_n<const N: usize>(
+        &mut self,
+    ) -> Result<(heapless::Vec<AccessPointInfo, N>, usize), EspError> {
+        // TODO: rewrite to do async waiting here while scanning
+        // Also, it would be nice to have an API that returned entries as they were discovered
+        self.base_driver.borrow_mut().scan_n()
+    }
+
+    pub async fn scan(&mut self) -> Result<Vec<AccessPointInfo>, EspError> {
+        let base: &mut WifiDriver<'d> = self.base_driver.borrow_mut();
+        base.scan()
+    }
+}
+
+#[cfg(all(feature = "nighty", feature = "experimental"))]
+impl<'d, T> asynch::Wifi for AsyncWifiDriver<'d, T>
+where
+    T: BorrowMut<WifiDriver<'d>>,
+{
+    type Error = EspError;
+
+    type GetCapabilitiesFuture<'a> = futures::future::Ready<Result<EnumSet<Capability>, EspError>> where Self: 'a;
+    type GetConfigurationFuture<'a> = futures::future::Ready<Result<Configuration, EspError>> where Self: 'a;
+    type SetConfigurationFuture<'a> = futures::future::Ready<Result<(), EspError>> where Self: 'a;
+    type StartFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<(), EspError>> + 'a>> where Self: 'a;
+    type StopFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<(), EspError>> + 'a>> where Self: 'a;
+    type ConnectFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<(), EspError>> + 'a>> where Self: 'a;
+    type DisconnectFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<(), EspError>> + 'a>> where Self: 'a;
+    type IsStartedFuture<'a> = futures::future::Ready<Result<bool, EspError>> where Self: 'a;
+    type IsConnectedFuture<'a> = futures::future::Ready<Result<bool, EspError>> where Self: 'a;
+    type ScanNFuture<'a, const N: usize> = Pin<Box<dyn futures::future::Future<Output = Result<(heapless::Vec<AccessPointInfo, N>, usize), EspError>> + 'a>> where Self: 'a;
+    type ScanFuture<'a> = Pin<Box<dyn futures::future::Future<Output = Result<alloc::vec::Vec<AccessPointInfo>, EspError>> + 'a>> where Self: 'a;
+
+    fn get_capabilities(&self) -> Self::GetCapabilitiesFuture<'_> {
+        futures::future::ready(self.get_capabilities())
+    }
+
+    fn get_configuration(&self) -> Self::GetConfigurationFuture<'_> {
+        futures::future::ready(self.get_configuration())
+    }
+
+    fn set_configuration(&mut self, conf: &Configuration) -> Self::SetConfigurationFuture<'_> {
+        futures::future::ready(self.set_configuration(conf))
+    }
+
+    fn start(&mut self) -> Self::StartFuture<'_> {
```

Comment:
Hint: and here you just need to do `async move { AsyncWifiDriver::start(self).await }`. In fact, even simply `AsyncWifiDriver::start(self)` without the async block and without the await should work too, but you should check.